mono/metadata/class-init.c

```diff
@@ -3151,60 +3151,58 @@ mono_class_setup_vtable_general (MonoClass *klass, MonoMethod **overrides, int o
+				if (mono_class_has_failure (klass))  /*Might be set by check_interface_method_override*/
+					goto fail;
+			}
+			
+			// If the slot is still empty, look in all the inherited virtual methods...
+			if ((vtable [im_slot] == NULL) && klass->parent != NULL) {
+				MonoClass *parent = klass->parent;
+				// Reverse order, so that last added methods are preferred
+				for (cm_index = parent->vtable_size - 1; cm_index >= 0; cm_index--) {
+					MonoMethod *cm = parent->vtable [cm_index];
+					
+					TRACE_INTERFACE_VTABLE ((cm != NULL) && printf ("    For slot %d ('%s'.'%s':'%s'), trying (ancestor) method '%s'.'%s':'%s'... ", im_slot, ic->name_space, ic->name, im->name, cm->klass->name_space, cm->klass->name, cm->name));
+					if ((cm != NULL) && check_interface_method_override (klass, im, cm, FALSE, FALSE, TRUE)) {
+						TRACE_INTERFACE_VTABLE (printf ("[everything ok]: ASSIGNING"));
 						vtable [im_slot] = cm;
 						/* Why do we need this? */
 						if (cm->slot < 0) {
 							cm->slot = im_slot;
 						}
+						if (conflict_map)
+							g_hash_table_remove(conflict_map, im);
+						break;
 					}
 					TRACE_INTERFACE_VTABLE (printf ("\n"));
 					if (mono_class_has_failure (klass))  /*Might be set by check_interface_method_override*/
 						goto fail;
 				}
-				
-				// If the slot is still empty, look in all the inherited virtual methods...
-				if ((vtable [im_slot] == NULL) && klass->parent != NULL) {
-					MonoClass *parent = klass->parent;
-					// Reverse order, so that last added methods are preferred
-					for (cm_index = parent->vtable_size - 1; cm_index >= 0; cm_index--) {
-						MonoMethod *cm = parent->vtable [cm_index];
-						
-						TRACE_INTERFACE_VTABLE ((cm != NULL) && printf ("    For slot %d ('%s'.'%s':'%s'), trying (ancestor) method '%s'.'%s':'%s'... ", im_slot, ic->name_space, ic->name, im->name, cm->klass->name_space, cm->klass->name, cm->name));
-						if ((cm != NULL) && check_interface_method_override (klass, im, cm, FALSE, FALSE, TRUE)) {
-							TRACE_INTERFACE_VTABLE (printf ("[everything ok]: ASSIGNING"));
-							vtable [im_slot] = cm;
-							/* Why do we need this? */
-							if (cm->slot < 0) {
-								cm->slot = im_slot;
-							}
-							break;
-						}
-						if (mono_class_has_failure (klass)) /*Might be set by check_interface_method_override*/
-							goto fail;
-						TRACE_INTERFACE_VTABLE ((cm != NULL) && printf ("\n"));
-					}
-				}
+			}
 
-				if (vtable [im_slot] == NULL) {
-					if (!(im->flags & METHOD_ATTRIBUTE_ABSTRACT)) {
-						TRACE_INTERFACE_VTABLE (printf ("    Using default iface method %s.\n", mono_method_full_name (im, 1)));
-						vtable [im_slot] = im;
-					}
+			if (vtable [im_slot] == NULL) {
+				if (!(im->flags & METHOD_ATTRIBUTE_ABSTRACT)) {
+					TRACE_INTERFACE_VTABLE (printf ("    Using default iface method %s.\n", mono_method_full_name (im, 1)));
+					vtable [im_slot] = im;
 				}
-			} else {
-				g_assert (vtable [im_slot] == override_im);
```

Comment:
Is there anything we can still assert at the end?  for example: either we found an override in the class (or its parent) itself or else the vtable slot is equal to the override_im?

Reply:
Yes, I can check if the vtable[im_slot] is equal cm that is the override in the class(or its parent) or is equal to override_im.

Reply:
Yes. Done.